### PR TITLE
EPH lease manager and checkpointing fixes and improvements

### PR DESCRIPTION
**Pull request by serkantkaraca**
Addressing issue https://github.com/Azure/azure-event-hubs-dotnet/issues/115

## Description
Here is the list of the changes:
  * Moving Offset and SequenceNumbre to base class Lease.cs
  * AzureStorageCheckpointLeaseManager.cs to stop downloading lease at every checkpoint call. This was causing 2 hosts operating on the same lease for a while.
  * Set MaximumExecutionTime on renewRequestOptions used while renewing storage lease.
  * Throw LeaseLostException when lease is lost on operations like Renew and Update.
  * Updating interface ICheckpointManager to accept Lease while updating checkpoint.
  * Some Linq improvements at lease steal and balancing logics.
  * Updating multiple-hosts unit test to validate load balancing.
  * Fixing an issue while checkpointing individual messages. Adding a unit test to cover this usage pattern.
  * ReceiveHandler and PartitionPump should gracefully handle exceptions thrown at ProcessError callbacks. Adding a CIT to cover error cases in EPH.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [X] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [X] If applicable, the public code is properly documented.
- [X] Pull request includes test coverage for the included changes.
- [X] The code builds without any errors.

### CODE REVIEW

**Comment by serkantkaraca** [comment]
Last commit addressing https://github.com/Azure/azure-event-hubs-dotnet/issues/116

File: src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs [hunk 1]
```diff
@@ -49,8 +49,9 @@ public interface ICheckpointManager
         /// <summary>
         /// Update the checkpoint in the store with the offset/sequenceNumber in the provided checkpoint.
         /// </summary>
+        /// <param name="lease">Partition information against which to perform a checkpoint.</param>
         /// <param name="checkpoint">offset/sequeceNumber to update the store with.</param>
-        Task UpdateCheckpointAsync(Checkpoint checkpoint);
+        Task UpdateCheckpointAsync(Lease lease, Checkpoint checkpoint);
```
Comment (jtaubensee) [inline src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs]: Is this a breaking change?

Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs]: Yes and no for .NetCore. Yes if anyone implemented their own checkpoint manager but probably not. The same change is also needed for Java. It is more dreadful there.

---
In reply to: [100396157](https://github.com/Azure/azure-event-hubs-dotnet/pull/117#discussion_r100396157) [](ancestors = 100396157)

Comment (jtaubensee) [inline src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs]: Can we deprecate this signature, and add the new one? That will allow us to release this a minor change according to SemVer.

File: src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs [hunk 2]
```diff
@@ -46,11 +46,15 @@ public interface ICheckpointManager
         /// <returns>The checkpoint for the given partition, whether newly created or already existing.</returns>
         Task<Checkpoint> CreateCheckpointIfNotExistsAsync(string partitionId);
 
+        [System.Obsolete("Use UpdateCheckpointAsync(Lease lease, Checkpoint checkpoint) instead", true)]
+        Task UpdateCheckpointAsync(Checkpoint checkpoint);
+
```
Comment (sjkwak) [inline src/Microsoft.Azure.EventHubs.Processor/ICheckpointManager.cs]: Let's remove this one.

File: src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs [hunk 3]
```diff
@@ -106,13 +112,18 @@ public Task<bool> CreateCheckpointStoreIfNotExistsAsync()
             return checkpoint;
         }
 
-        public async Task UpdateCheckpointAsync(Checkpoint checkpoint)
+        [Obsolete("Use UpdateCheckpointAsync(Lease lease, Checkpoint checkpoint) instead", true)]
+        public Task UpdateCheckpointAsync(Checkpoint checkpoint)
+        {
+            throw new NotImplementedException();
+        }
```
Comment (sjkwak) [inline src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs]: Let's remove this one.

File: src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs [hunk 4]
```diff
@@ -106,13 +112,18 @@ public Task<bool> CreateCheckpointStoreIfNotExistsAsync()
...
+
+        public async Task UpdateCheckpointAsync(Lease lease, Checkpoint checkpoint)
         {
-            // Need to fetch the most current lease data so that we can update it correctly.
-            AzureBlobLease lease = (AzureBlobLease)await GetLeaseAsync(checkpoint.PartitionId).ConfigureAwait(false);
-            lease.Offset = checkpoint.Offset;
-            lease.SequenceNumber = checkpoint.SequenceNumber;
-            await UpdateLeaseAsync(lease).ConfigureAwait(false);
+            AzureBlobLease newLease = new AzureBlobLease((AzureBlobLease)lease);
```
Comment (sjkwak) [inline src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs]: >zureBlobLease newLease = new AzureBlobLease((AzureBlobLease)lease); [](start = 13, length = 67)

we don't need cast here.

File: src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs [hunk 5]
```diff
@@ -106,13 +112,18 @@ public Task<bool> CreateCheckpointStoreIfNotExistsAsync()
...
+            newLease.Offset = checkpoint.Offset;
+            newLease.SequenceNumber = checkpoint.SequenceNumber;
+            await UpdateLeaseAsync(newLease).ConfigureAwait(false);
```
Comment (sjkwak) [inline src/Microsoft.Azure.EventHubs.Processor/AzureStorageCheckpointLeaseManager.cs]: >UpdateLeaseAsync [](start = 18, length = 16)

this.

**Pull request merged**